validator_client/src/config.rs

```diff
@@ -1,32 +1,131 @@
+    /// The server at which the Beacon Node can be contacted
     pub server: String,
+    /// The chain specification that we are connecting to
     pub spec: ChainSpec,
 }
 
-const DEFAULT_LIGHTHOUSE_DIR: &str = ".lighthouse-validators";
+const DEFAULT_VALIDATOR_DATADIR: &str = ".lighthouse-validator";
+const DEFAULT_VALIDATORS_SUBDIR: &str = "validators";
+const DEFAULT_PRIVATE_KEY_FILENAME: &str = "private.key";
 
-impl ClientConfig {
-    /// Build a new configuration from defaults.
-    pub fn default() -> Self {
-        let data_dir = {
-            let home = dirs::home_dir().expect("Unable to determine home dir.");
-            home.join(DEFAULT_LIGHTHOUSE_DIR)
+impl ValidatorClientConfig {
+    /// Build a new configuration from defaults, which are overrided by arguments provided.
+    pub fn build_config(arguments: &ArgMatches) -> Result<Self, Error> {
+        // Use the specified datadir, or default in the home directory
+        let data_dir: PathBuf = match arguments.value_of("datadir") {
+            Some(path) => PathBuf::from(path.to_string()),
+            None => {
+                let home = dirs::home_dir().ok_or_else(|| {
+                    Error::new(ErrorKind::NotFound, "Unable to determine home directory.")
+                })?;
+                home.join(DEFAULT_VALIDATOR_DATADIR)
+            }
         };
-        fs::create_dir_all(&data_dir)
-            .unwrap_or_else(|_| panic!("Unable to create {:?}", &data_dir));
-        let server = "localhost:50051".to_string();
-        let spec = ChainSpec::foundation();
-        Self {
+        fs::create_dir_all(&data_dir)?;
+
+        let validator_dir = data_dir.join(DEFAULT_VALIDATORS_SUBDIR);
+        fs::create_dir_all(&validator_dir)?;
+
+        let server: String = match arguments.value_of("server") {
+            Some(srv) => {
+                //TODO: I don't think this parses correctly a server & port combo
+                srv.parse::<u16>()
+                    .map_err(|e| Error::new(ErrorKind::InvalidInput, e))?
+                    .to_string()
+            }
+            None => "localhost:50051".to_string(),
+        };
+
+        // TODO: Permit loading a custom spec from file.
+        let spec: ChainSpec = match arguments.value_of("spec") {
+            Some(spec_str) => {
+                match spec_str {
+                    "foundation" => ChainSpec::foundation(),
+                    "few_validators" => ChainSpec::few_validators(),
+                    // Should be impossible due to clap's `possible_values(..)` function.
+                    _ => unreachable!(),
+                }
+            }
+            None => ChainSpec::foundation(),
+        };
+
+        Ok(Self {
             data_dir,
+            validator_dir,
             server,
             spec,
+        })
+    }
+
+    /// Try to load keys from validator_dir, returning None if none are found or an error.
+    pub fn fetch_keys(&self) -> Result<Option<Vec<Keypair>>, Error> {
```

Comment:
In my early days, I was using a lot of `Result<Option<>>` as well. Later I realise I don't do it that much. It may be applicable here, but if we returned an empty array, would that also be fine?